cmd/catalog-importer/cmd/sync.go

```diff
@@ -240,33 +240,63 @@ func (opt *SyncOptions) Run(ctx context.Context, logger kitlog.Logger, cfg *conf
 		}
 	}
 
-	// Update type schemas to match config
 	OUT("\n↻ Syncing catalog type schemas...")
-	for _, outputType := range cfg.Outputs() {
-		baseModel, enumModels := output.MarshalType(outputType)
-		for _, model := range append(enumModels, baseModel) {
-			catalogType := catalogTypesByOutput[model.TypeName]
-
-			var updatedCatalogType client.CatalogTypeV2
-			if opt.DryRun {
-				logger.Log("msg", "dry-run active, which means we fake a response")
-				updatedCatalogType = *catalogType // they start the same
-
-				// Then we pretend like we've already updated the schema, which means we rebuild the
-				// attributes.
-				updatedCatalogType.Schema = client.CatalogTypeSchemaV2{
-					Version:    updatedCatalogType.Schema.Version,
-					Attributes: []client.CatalogTypeAttributeV2{},
+	if opt.DryRun {
+		for _, outputType := range cfg.Outputs() {
+			baseModel, enumModels := output.MarshalType(outputType)
+			for _, model := range append(enumModels, baseModel) {
+				catalogType := catalogTypesByOutput[model.TypeName]
+
+				var updatedCatalogType client.CatalogTypeV2
+				if opt.DryRun {
+					logger.Log("msg", "dry-run active, which means we fake a response")
+					updatedCatalogType = *catalogType // they start the same
+
+					// Then we pretend like we've already updated the schema, which means we rebuild the
+					// attributes.
+					updatedCatalogType.Schema = client.CatalogTypeSchemaV2{
+						Version:    updatedCatalogType.Schema.Version,
+						Attributes: []client.CatalogTypeAttributeV2{},
+					}
+					for _, attr := range model.Attributes {
+						updatedCatalogType.Schema.Attributes = append(updatedCatalogType.Schema.Attributes, client.CatalogTypeAttributeV2{
+							Id:                *attr.Id,
+							Name:              attr.Name,
+							Type:              attr.Type,
+							Array:             attr.Array,
+							Mode:              client.CatalogTypeAttributeV2Mode(*attr.Mode),
+							BacklinkAttribute: attr.BacklinkAttribute,
+						})
+					}
 				}
+				OUT("  ✔ %s (id=%s)", model.TypeName, catalogType.Id)
+				DIFF("  ", *catalogType, updatedCatalogType)
+			}
+		}
+	} else {
+		// Update all the type schemas except for new backlinks, which could reference
+		// attributes that don't exist yet.
+		catalogTypeVersions := map[string]int64{}
+		for _, outputType := range cfg.Outputs() {
```

Comment:
I don't think this is a performance problem that needs to be fixed, but we're going through the types twice - once to figure out what's a new backlink, and once to figure out what isn't.

Using `lo.PartitionBy` might be useful here to do this in one go, then we only have to implement one bit of logic that works out if an attribute is a new backlink or not.

Reply:
Yeah I felt that writing the code like that was kinda more confusing, and given #attributes should be tiny, I think this is fine?
Main thing that would change my mind if it this was confusing to you as a reader